client/src/components/ContentNavigator/ContentDataProvider.ts

```diff
@@ -0,0 +1,179 @@
+            title: "Open SAS File",
+          },
+    };
+  }
+
+  public getChildren(item?: ContentItem): ProviderResult<ContentItem[]> {
+    return this.model.getChildren(item);
+  }
+
+  public watch(): Disposable {
+    // ignore, fires for all changes...
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    return new Disposable(() => {});
+  }
+
+  public async stat(uri: Uri): Promise<FileStat> {
+    return await this.model.getResourceByUri(uri).then(
+      (resource): FileStat => ({
+        type: this.dataDescriptor.isContainer(resource)
+          ? FileType.Directory
+          : FileType.File,
+        ctime: this.dataDescriptor.getCreationDate(resource),
+        mtime: this.dataDescriptor.getModifyDate(resource),
+        size: 0,
+      })
+    );
+  }
+
+  public async readFile(uri: Uri): Promise<Uint8Array> {
+    return await this.model
+      .getContentByUri(uri)
+      .then((content) => new TextEncoder().encode(content));
+  }
+
+  public getUri(item: ContentItem): Promise<Uri> {
+    return this.model.getUri(item);
+  }
+
+  public async createFolder(
+    item: ContentItem,
+    folderName: string
+  ): Promise<Uri | undefined> {
+    const newItem = await this.model.createFolder(item, folderName);
+    if (newItem) {
+      this.refresh();
+      return this.dataDescriptor.getUri(newItem);
+    }
+  }
+
+  public async createFile(
+    item: ContentItem,
+    fileName: string
+  ): Promise<Uri | undefined> {
+    const newItem = await this.model.createFile(item, fileName);
+    if (newItem) {
+      this.refresh();
+      return this.dataDescriptor.getUri(newItem);
+    }
+  }
+
+  public async renameResource(
+    item: ContentItem,
+    name: string
+  ): Promise<Uri | undefined> {
+    const newItem = await this.model.renameResource(item, name);
+    if (newItem) {
+      return this.dataDescriptor.getUri(newItem);
+    }
+  }
+
+  public writeFile(uri: Uri, content: Uint8Array): void | Promise<void> {
+    return this.model.saveContentToUri(uri, new TextDecoder().decode(content));
+  }
+
+  public async deleteResource(item: ContentItem): Promise<boolean> {
+    const success = await this.model.delete(item);
+    if (success) {
+      this.refresh();
+    }
+
+    return success;
+  }
+
+  public refresh(): void {
+    this._onDidChangeTreeData.fire(undefined);
+  }
+
+  public async getParent(
+    element: ContentItem
+  ): Promise<ContentItem | undefined> {
+    return await this.model.getParent(element);
+  }
+
+  public async delete(): Promise<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  public rename(): void | Promise<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  public readDirectory():
+    | [string, FileType][]
+    | Thenable<[string, FileType][]> {
+    throw new Error("Method not implemented.");
+  }
+
+  public createDirectory(): void | Thenable<void> {
+    throw new Error("Method not implemented.");
+  }
+}
```

Comment:
These are the methods that aren't implemented:
 - `delete`
 - `rename`
 - `readDirectory`
 - `createDirectory`

Fwiw, `readDirectory` and `createDirectory` were never triggered in any testing. Thus, those were left unimplemented.

Both `delete` and `rename` were a bit hard to work with in the context of a tree data provider (instead of purely a `FileSystemProvider`). Thus,

Reply:
For example, we _could_ implement `delete`. It would involve:
 - Getting the uri of the `ContentItem` when calling delete from the command
 - Calling `getResourceByUri`. This will give us _back_ the ContentItem we already had
 - Finding the delete link and using that to delete the entry

For that one, I think sticking with our custom implementation may make the most sense, especially as we start implementing "recycle bin" functionality where "delete" is a more nuanced concept

Reply:
Got it. It makes sense to me.